source/extensions/filters/network/thrift_proxy/protocol.h

```diff
@@ -78,6 +78,48 @@ enum class FieldType {
   LastFieldType = List,
 };
 
+/**
+ * ProtocolCallbacks are Thrift protocol-level callbacks.
+ */
+class ProtocolCallbacks {
+public:
+  virtual ~ProtocolCallbacks() {}
+
+  /**
+   * Indicates that the start of a Thrift protocol message was detected.
+   * @param name the name of the message, if available
+   * @param msg_type the type of the message
+   * @param seq_id the message sequence id
+   */
+  virtual void messageStart(const absl::string_view name, MessageType msg_type,
+                            int32_t seq_id) PURE;
+
+  /**
+   * Indicates that the start of a Thrift protocol struct was detected.
+   * @param name the name of the struct, if available
+   */
+  virtual void structBegin(const absl::string_view name) PURE;
+
+  /**
+   * Indicates that a Thrift protocol struct field was detected.
```

Comment:
This doesn't specify whether the callback is invoked at the beginning or end of the field, although the implementation in this PR calls the callback at the beginning. For the use cases where you're planning to implement `ProtocolCallbacks`, will that distinction matter?

Reply:
The distinction doesn't matter at this point -- we'll use this to detect whether the outermost struct of a reply contains field id 0 (indicating success irrespective of the field value) or not (IDL-level exception). I'll update it to make it clear, though.